### PR TITLE
Support building and publishing multiple versions of specific package as part of package repository and updated tkg-autoscaler package to use multiple versions

### LINKED ISSUE #2580: Support building multiple packages of different version and publish it as part of package repository
(This is used to request new product features)

**Describe the feature request**

Currently, when building packages, we can only have 1 version of a specific package part of the package repository we build. We would like to support publishing multiple versions of the same package to the package repository.

For example, we have `tkg-autoscaler` package that needs to have 3 version of that package (v1.21.0, v1.22.0, 1.23.0) to be part of management package repository. Currently, we cannot add only 1 version of the specific package to `package-values.yaml` and the build/publish scripts only supports 1 version of a specific plugin.

Enhance the build/publish scripts to support building and publishing multiple version of a single package as part of package repository.

**Describe alternatives you've considered**
NA

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [x] Test and Release
- [ ] User Experience

**Additional context**

### PR DESCRIPTION
**Pull request by anujc25**
### What this PR does / why we need it
- Support building multiple packages of different version
- Updated `tkg-autoscaler` package to support building different version of same package with different autoscaler image

### Which issue(s) this PR fixes

Fixes #2580

#### Building package repository and deploying the package repository 
1. Build and publish all management packages
```
$ export OCI_REGISTRY=gcr.io/eminent-nation-87317/tkg/test24
$ make package-push-bundles-repo PACKAGE_REPOSITORY=management 
```
3. Create a bootstrap kind cluster and install kapp-controller to the cluster
4. Install management package repository to the cluster
```
$ tanzu package repository update tanzu-management --create --url gcr.io/eminent-nation-87317/tkg/test24/management:v0.23.0-dev
```
5. Verify the installed packages on the cluster.
```
~ $ kubectl get package -A
NAMESPACE   NAME                                                      PACKAGEMETADATA NAME                           VERSION      AGE
default     addons-manager.tanzu.vmware.com.0.23.0-dev                addons-manager.tanzu.vmware.com                0.23.0-dev   3s
default     cliplugins.tanzu.vmware.com.0.23.0-dev                    cliplugins.tanzu.vmware.com                    0.23.0-dev   3s
default     featuregates.tanzu.vmware.com.0.23.0-dev                  featuregates.tanzu.vmware.com                  0.23.0-dev   3s
default     framework.tanzu.vmware.com.0.23.0-dev                     framework.tanzu.vmware.com                     0.23.0-dev   3s
default     tanzu-auth.tanzu.vmware.com.0.23.0-dev                    tanzu-auth.tanzu.vmware.com                    0.23.0-dev   3s
default     tkg-autoscaler.tanzu.vmware.com.1.21.0                    tkg-autoscaler.tanzu.vmware.com                1.21.0       3s
default     tkg-autoscaler.tanzu.vmware.com.1.22.0                    tkg-autoscaler.tanzu.vmware.com                1.22.0       3s
default     tkg-autoscaler.tanzu.vmware.com.1.23.0                    tkg-autoscaler.tanzu.vmware.com                1.23.0       3s
default     tkg-clusterclass-aws.tanzu.vmware.com.0.23.0-dev          tkg-clusterclass-aws.tanzu.vmware.com          0.23.0-dev   3s
default     tkg-clusterclass-azure.tanzu.vmware.com.0.23.0-dev        tkg-clusterclass-azure.tanzu.vmware.com        0.23.0-dev   3s
default     tkg-clusterclass-docker.tanzu.vmware.com.0.23.0-dev       tkg-clusterclass-docker.tanzu.vmware.com       0.23.0-dev   3s
default     tkg-clusterclass-vsphere.tanzu.vmware.com.0.23.0-dev      tkg-clusterclass-vsphere.tanzu.vmware.com      0.23.0-dev   3s
default     tkg-clusterclass.tanzu.vmware.com.0.23.0-dev              tkg-clusterclass.tanzu.vmware.com              0.23.0-dev   3s
default     tkg.tanzu.vmware.com.0.23.0-dev                           tkg.tanzu.vmware.com                           0.23.0-dev   3s
default     tkr-aws-machine-webhook.tanzu.vmware.com.0.23.0-dev       tkr-aws-machine-webhook.tanzu.vmware.com       0.23.0-dev   3s
default     tkr-azure-machine-webhook.tanzu.vmware.com.0.23.0-dev     tkr-azure-machine-webhook.tanzu.vmware.com     0.23.0-dev   3s
default     tkr-docker-machine-webhook.tanzu.vmware.com.0.23.0-dev    tkr-docker-machine-webhook.tanzu.vmware.com    0.23.0-dev   3s
default     tkr-service.tanzu.vmware.com.0.23.0-dev                   tkr-service.tanzu.vmware.com                   0.23.0-dev   3s
default     tkr-source-controller.tanzu.vmware.com.0.23.0-dev         tkr-source-controller.tanzu.vmware.com         0.23.0-dev   3s
default     tkr-vsphere-machine-webhook.tanzu.vmware.com.0.23.0-dev   tkr-vsphere-machine-webhook.tanzu.vmware.com   0.23.0-dev   3s
```
6. Verified the output has 3 different versions of tkg-autoscaler package.

#### Building a single package
```
$ make package-bundle PACKAGE_NAME=tkg-autoscaler
.
.
cd hack/packages/package-tools && go run main.go package-bundle generate tkg-autoscaler --thick --version=v0.23.0-dev --sub-version= --registry=gcr.io/eminent-nation-87317/tkg/test25
Generating "tkg-autoscaler" package bundle...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.21.0'...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.22.0'...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.23.0'...
```
```
$ make package-bundle PACKAGE_NAME=tkg-autoscaler PACKAGE_VERSION=v1.22.0
.
.
cd hack/packages/package-tools && go run main.go package-bundle generate tkg-autoscaler --thick --version=v0.23.0-dev --sub-version= --registry=gcr.io/eminent-nation-87317/tkg/test25
Generating "tkg-autoscaler" package bundle...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.22.0'...
```

### Release note

```release-note
Support building and publishing multiple versions of specific package as part of package repository
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Comment by vuil** [comment]
Thanks, that's a nice piece of enhancement!

Might be good to file an feature request issue to document the problem being solved.
This can use a short release note as it gives more flexibility to package publishing process
> Verified the output has 2 different versions
I think you have 3 :-)
There are some linter nits.

**Comment by anujc25** [comment]
> i think we should also generate package bundles for all the versions for a package when we run make package-bundle PACKAGE_NAME=foo, right now with this enhancement package bundles of all the versions of the package are only generated when we run make package-bundles

Thanks @yharish991 for pointing out. I tested the `make package-bundle` and it is working fine. However, I added another commit on top of this to support one more case of "Support building a single package by providing specific version when there are multiple packages exists with same name in package-values.yaml"

For example:
- `package-values.yaml` contains 3 `tkg-autoscaler` packages with different version.
- User can now run `make package-bundle PACKAGE_NAME=tkg-autoscaler PACKAGE_VERSION=v1.23.0` to build `v1.23.0` version specific package among 3 pacakges.

Tested this with:
```
$ make package-bundle PACKAGE_NAME=tkg-autoscaler PACKAGE_VERSION=v1.23.0 # builds v1.23.0 autoscaler package
$ make package-bundle PACKAGE_NAME=tkg-autoscaler # builds first available v1.21.0 autoscaler package
$ make package-bundle PACKAGE_NAME=addons-manager # works as before
```

**Comment by rajathagasthya** [comment]
> $ make package-bundle PACKAGE_NAME=tkg-autoscaler # builds all 3 autoscaler packages v1.21.0, v1.22.0, v1.23.0

I understand why this change was made, but this is not immediately obvious if you don't already know what it does. I don't have an alternative though.

**Comment by anujc25** [comment]
> I understand why this change was made, but this is not immediately obvious if you don't already know what it does. I don't have an alternative though.

To build a specific version user can always pass `PACKAGE_VERSION` with the command.

I couldn't find any other way to make it obvious to the user except updating the logs that we show when we build the packages. I have updated the log and it will show that we are building all 3 autoscaler packages like below:
```
$ make package-bundle PACKAGE_NAME=tkg-autoscaler
.
.
cd hack/packages/package-tools && go run main.go package-bundle generate tkg-autoscaler --thick --version=v0.23.0-dev --sub-version= --registry=gcr.io/eminent-nation-87317/tkg/test25
Generating "tkg-autoscaler" package bundle...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.21.0'...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.22.0'...
Including thick tarball...
Generating Package CR for package 'tkg-autoscaler:v1.23.0'...
```

**Comment by vuil** [comment]
changes look reasonable to me, but I will let @yharish991 and @rajathagasthya sign off on this one.

**Comment by yharish991** [comment]
I'm fine with the approach, will do another round of review by EOD..

File: packages/package-values.yaml [hunk 1]
```diff
@@ -72,14 +72,41 @@ repositories:
         version: latest
         #! Gets replaced with imgpkg sha256 at build, this should be name:version
         sha256: "tkg-clusterclass:latest"
+
```
Comment (yharish991) [inline packages/package-values.yaml:75]: nit: can we remove this line?

File: packages/package-values.yaml [hunk 2]
```diff
@@ -72,14 +72,41 @@ repositories:
...
+        #! Relative path to package bundle
+        path: packages/tkg-autoscaler
+        domain: tanzu.vmware.com
+        version: v1.23.0
+        #! Gets replaced with imgpkg sha256 at build, this should be name:version
+        sha256: "tkg-autoscaler:latest"
+        skipVersionOverride: true
+        env:
+          AUTOSCALER_IMAGE: projects-stg.registry.vmware.com\/tkg\/cluster-autoscaler:v1.23.0_vmware.1
+
```
Comment (yharish991) [inline packages/package-values.yaml:109]: nit: same here, can you remove the empty line?

**Review by yharish991 — APPROVED** [review]
LGTM, added very minor nits..

**Review by rajathagasthya — APPROVED** [review]

**Pull request merged**